internal/util/regexp.go

```diff
@@ -0,0 +1,43 @@
+// Copyright 2021-2023 Zenauth Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+package util
+
+import (
+	"fmt"
+	"regexp"
+	"sync"
+)
+
+type RegexpCache struct {
+	cache map[string]*regexp.Regexp
+	mu    *sync.Mutex
```

Comment:
I would use a `RWMutex` here 

```suggestion
	mu    sync.RWMutex
```